week3/RESTfulAPIs/app.js

```diff
@@ -0,0 +1,136 @@
+//start promise
+const readTodosFile = (filename) => {
+    return new Promise((resolve, reject) => {
+        fs.readFile(filename, "utf8", (err, data) => {
+            if (err) {
+                if (err.code === "ENOENT") {
+                    console.log("Sorry, this file doesn't exist.");
+                } else if (err.code === "EACCES") {
+                    console.log("Sorry, but you don't have permission.");
+                } else {
+                    console.log(` unknown error occurred: ${err.message}`);
+                }
+                reject(err);
+            }
+            console.log(`Successfully read the file ${filename}`);
+            resolve(data);
+        });
+    });
+};
+const writeTodosFile = (filename, dataAsString) => {
+    return new Promise((resolve, reject) => {
+        fs.writeFile(filename, dataAsString, (err) => {
+            if (err) {
+                reject(err);
+            }
+            resolve(`The < ${filename}> file has been saved!`);
+        });
+    });
+};
+
+//add task
+const addTodoItem = (filename, todoList, todoDescription) => {
+    todoList.push({
+        "description": todoDescription,
+        "id": uniqid()
+    });
+    writeTodosFile(TODOS_FILENAME, JSON.stringify(myTodoList))
+        .then(() => console.log("Done adding todo."))
+        .catch((err) => {
+            console.log('ERROR:', err);
+        });
+};
+
+const removeTodoItem = (filename, todoList, todoId) => {
+    const index = todoList.map((x) => {
+        return x.Id;
+    }).indexOf(todoId);
+    todoList.splice(index, 1);
+
+    todoList.filter(e => e.id !== todoId);
+    writeTodosFile(TODOS_FILENAME, JSON.stringify(myTodoList))
+        .then(() => console.log("Done Removing the task from your TODO list."))
+        .catch((err) => {
+            console.log('ERROR:', err);
+        });
+};
+
+const updateTodoItem = (filename, todoList, newDescription, todoId) => {
+    for (const i in todoList) {
+        if (todoList[i].id === todoId) {
+            todoList[i].description = newDescription;
+            break;
+        } 
+    }
+    writeTodosFile(TODOS_FILENAME, JSON.stringify(myTodoList))
+        .then(() => console.log("You task has been updated ."))
+        .catch((err) => {
+            console.log('ERROR:', err);
+        });
+};
+
+readTodosFile(TODOS_FILENAME)
+    .then((data) => {
+        data = data || "[]";
+        myTodoList = JSON.parse(data);
+
+        app.get('/', (req, res) => {
+            readTodosFile('./help/help.txt').then((data) => {
+                res.send(data);
+            }).catch((err) => console.log(err));
+        });
+
+        app.get('/list', (req, res) => {
+            res.send(myTodoList);
+        });
+
+        app.post('/add/:description', (req, res) => {
+            const description = req.params.description;
+            addTodoItem(TODOS_FILENAME, myTodoList, description);
+            res.send('added Task done..');
```

Comment:
So - because `addTodoItem` is async (inside, it calls a function that returns a promise), it's possible that there's a problem or error which means the task cannot be added. But since you don't handle async cases here, you will `res.send()` before the task is done being added. That means if there's a problem, you cannot notify the 'user' who sent this request - you already told them it was done before you had finished trying!

You probably want to refactor this so that `addTodoItem` returns a promise, and then change this to:
```
addTodoItem(TODOS_FILENAME, myTodoList, description).then( () => {
  res.send('added Task done..')
}).catch( () => {
  // error handling... 
});
```